lib/charms/opensearch/v0/opensearch_snapshots.py

```diff
@@ -0,0 +1,1628 @@
+        self.framework.observe(self.charm.on.restore_action, self._on_restore_action)
+
+    def _on_s3_credentials_changed(self, event: CredentialsChangedEvent) -> None:  # noqa: C901
+        """Handler for s3 credentials changed event."""
+        dep = self.charm.opensearch_peer_cm.deployment_desc()
+        # block non-main orchestrators only when they are in a multi-app topology.
+        if (
+            dep
+            and dep.typ != DeploymentType.MAIN_ORCHESTRATOR
+            and self._is_part_of_large_deployment()
+        ):
+            if self.charm.unit.is_leader():
+                self.charm.status.set(BlockedStatus(BackupRelShouldNotExist), app=True)
+                self.charm.status.set(BlockedStatus(BackupRelShouldNotExist))
+            return
+
+        object_storage_type = self.charm.snapshots_manager.resolver.get_storage_type()
+        logger.info(f"S3 credentials changed for object storage type {object_storage_type}")
+        if object_storage_type == "conflict":
+            if self.charm.unit.is_leader():
+                self.charm.status.set(BlockedStatus(BackupRelConflict), app=True)
+                self.charm.status.set(BlockedStatus(BackupRelConflict))
+            event.defer()
+            return
+
+        if self.charm.unit.is_leader():
+            self.charm.status.clear(BackupRelConflict, app=True)
+            self.charm.status.clear(BackupRelConflict)
+
+        cfg = self.charm.snapshots_manager.resolver.get_storage_config(ObjectStorageType.S3)
+        creds = getattr(getattr(cfg, "s3", None), "credentials", None)
+        # handle case where this was deferred in the above case, then the s3 relation was severed
+        if object_storage_type != "s3" or not cfg or not creds:
+            logger.warning("No S3 object storage configuration.")
+            if self.charm.unit.is_leader():
+                self.charm.status.set(BlockedStatus(BackupRelDataIncomplete), app=True)
+                self.charm.status.set(BlockedStatus(BackupRelDataIncomplete))
+            return
+
+        if self.charm.unit.is_leader():
+            self.charm.status.clear(BackupRelDataIncomplete, app=True)
+            self.charm.status.clear(BackupRelDataIncomplete)
+
+        # apply locally (leader does cluster-level config)
+        self.charm.keystore_manager.put_entries(
+            {
+                "s3.client.default.access_key": cfg.s3.credentials.access_key,
+                "s3.client.default.secret_key": cfg.s3.credentials.secret_key,
+            }
+        )
+        logger.info("S3 credentials are added to keystore.")
+        self.charm.keystore_manager.reload()
+
+        need_custom_ca = self.charm.snapshots_manager.requires_custom_s3_ca(
+            object_storage_type, cfg
+        )
+        new_chain = cfg.s3.tls_ca_chain if need_custom_ca else None
+
+        if new_chain:
+            self.charm.snapshots_manager.store_s3_ca(new_chain)
+            logger.info("S3 CA stored.")
```

Comment:
```suggestion
        if needs_custom_ca:
            self.charm.snapshots_manager.store_s3_ca(cfg.s3.tls_ca_chain)
            logger.info("S3 CA stored.")
```

Question: how do we handle when the CA chain is the same as the one currently configured with (if any) - on subsequent credential changes?

Reply:
If we get a new chain which aimed to stored, we run `OpenSearchSnapshotsManager.is_custom_s3_ca_stored(new_chain)` to compare the hashes of existing and the new CA. If they are same, no operation is done. We only store it if the below code returns false.

```
current_chain = self.find_s3_chain_in_store()
if not current_chain:
    return False

if not s3_ca_chain:
    # There is existing S3 CA stored, but no new one, we need to remove the old one.
    return True

stored_blocks = _normalize_chain_unordered(current_chain)
new_blocks = _normalize_chain_unordered(s3_ca_chain)

return stored_blocks == new_blocks
```